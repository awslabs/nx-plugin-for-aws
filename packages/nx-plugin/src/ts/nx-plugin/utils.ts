/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  joinPathFragments,
  readJson,
  Tree,
  updateJson,
  writeJson,
} from '@nx/devkit';
import { readProjectConfigurationUnqualified } from '../../utils/nx';
import { withVersions } from '../../utils/versions';
import PackageJson from '../../../package.json';

/**
 * Configures a TypeScript project as an Nx Plugin
 */
export const configureTsProjectAsNxPlugin = (
  tree: Tree,
  projectName: string,
) => {
  const project = readProjectConfigurationUnqualified(tree, projectName);

  const tsConfigPath = joinPathFragments(project.root, 'tsconfig.json');
  if (!tree.exists(tsConfigPath)) {
    throw new Error(
      `Selected plugin project ${projectName} is not a TypeScript project`,
    );
  }

  // Nx Plugins must use commonjs as a limitation of nx
  // https://github.com/nrwl/nx/issues/15682
  updateJson(tree, tsConfigPath, (tsConfig) => {
    tsConfig.compilerOptions ??= {};
    tsConfig.compilerOptions.module ??= 'commonjs';
    return tsConfig;
  });

  // Create an empty generators.json if one dosn't exist
  const generatorsJsonPath = joinPathFragments(project.root, 'generators.json');
  if (!tree.exists(generatorsJsonPath)) {
    writeJson(tree, generatorsJsonPath, {
      generators: {},
    });
  }

  // Create a package.json if one doesn't exist, and configure "main" and "generators"
  const pluginPackageJsonPath = joinPathFragments(project.root, 'package.json');
  if (!tree.exists(pluginPackageJsonPath)) {
    writeJson(tree, pluginPackageJsonPath, {
      name: project.name,
    });
  }
  updateJson(tree, pluginPackageJsonPath, (pkg) => {
    pkg.main ??= './src/index.js';
    pkg.generators ??= './generators.json';
    return pkg;
  });

  const rootPackageJson = tree.exists('package.json')
    ? readJson(tree, 'package.json')
    : undefined;
  const isNxPluginForAws = rootPackageJson?.name === '@aws/nx-plugin-source';

  if (!isNxPluginForAws) {
    // Add the required dependencies to the root package.json, and project's package.json
    const deps = {
      ...withVersions(['@nx/devkit']),
      // Include dependency on @aws/nx-plugin
      [PackageJson.name]: `^${PackageJson.version}`,
    };
    addDependenciesToPackageJson(tree, {}, deps);
    addDependenciesToPackageJson(tree, deps, {}, pluginPackageJsonPath);
  }
};
