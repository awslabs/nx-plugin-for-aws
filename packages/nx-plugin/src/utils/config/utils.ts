/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { generateFiles, joinPathFragments, Tree } from '@nx/devkit';
import { AwsNxPluginConfig } from '.';
import * as ts from 'typescript';
import { jsonToAst, replace } from '../ast';
import { factory } from 'typescript';
import { formatFilesInSubtree } from '../format';
import { importTypeScriptModule } from '../js';

export const AWS_NX_PLUGIN_CONFIG_FILE_NAME = 'aws-nx-plugin.config.mts';

/**
 * Ensure that the config file exists
 */
export const ensureAwsNxPluginConfig = async (
  tree: Tree,
): Promise<AwsNxPluginConfig> => {
  if (!tree.exists(AWS_NX_PLUGIN_CONFIG_FILE_NAME)) {
    // Create an empty config file if it doesn't already exist
    generateFiles(tree, joinPathFragments(__dirname, 'files'), '.', {});
  }
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return await readAwsNxPluginConfig(tree)!;
};

/**
 * Read config from the aws nx plugin configuration file
 */
export const readAwsNxPluginConfig = async (
  tree: Tree,
): Promise<AwsNxPluginConfig | undefined> => {
  if (!tree.exists(AWS_NX_PLUGIN_CONFIG_FILE_NAME)) {
    return undefined;
  }
  const configTs = tree.read(AWS_NX_PLUGIN_CONFIG_FILE_NAME, 'utf-8');
  return await importTypeScriptModule(configTs);
};

/**
 * Update the aws nx plugin config file.
 * Undefined top level keys in the config update will be untouched, otherwise config is replaced
 */
export const updateAwsNxPluginConfig = async (
  tree: Tree,
  configUpdate: Partial<AwsNxPluginConfig>,
): Promise<void> => {
  // Replace the default export
  replace(
    tree,
    AWS_NX_PLUGIN_CONFIG_FILE_NAME,
    'ExportAssignment ObjectLiteralExpression',
    (node) => {
      const existingObj = node as ts.ObjectLiteralExpression;

      const existingProps = new Map<string, ts.PropertyAssignment>();
      existingObj.properties.forEach((prop) => {
        if (ts.isPropertyAssignment(prop)) {
          existingProps.set(prop.name.getText(), prop);
        }
      });

      const properties: ts.PropertyAssignment[] = [];

      for (const [key, value] of Object.entries(configUpdate)) {
        properties.push(
          factory.createPropertyAssignment(
            key,
            jsonToAst(value) as ts.Expression,
          ),
        );
      }

      existingObj.properties.forEach((prop) => {
        if (ts.isPropertyAssignment(prop)) {
          const name = prop.name.getText();
          if (configUpdate[name] === undefined) {
            properties.push(prop);
          }
        }
      });

      return factory.createObjectLiteralExpression(properties, true);
    },
  );

  // Format the config nicely after an update
  await formatFilesInSubtree(tree, AWS_NX_PLUGIN_CONFIG_FILE_NAME);
};
