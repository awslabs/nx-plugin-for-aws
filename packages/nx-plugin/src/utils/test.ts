/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { vi, expect } from 'vitest';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { joinPathFragments, Tree } from '@nx/devkit';
import solutionSetup from '@nx/js/src/utils/typescript/ts-solution-setup';

export const createTreeUsingTsSolutionSetup = (): Tree => {
  vi.spyOn(solutionSetup, 'isUsingTsSolutionSetup').mockImplementation(
    () => true,
  );

  const tree = createTreeWithEmptyWorkspace();

  tree.write('pnpm-workspace.yaml', `packages:\n  - packages/*`);

  tree.write('tsconfig.json', '{}');
  return tree;
};

/**
 * Snapshot all files within a directory in the given tree
 */
export const snapshotTreeDir = (tree: Tree, dir: string) => {
  if (tree.isFile(dir)) {
    expect(tree.read(dir, 'utf-8')).toMatchSnapshot(dir);
  } else {
    tree
      .children(dir)
      .forEach((subDir) =>
        snapshotTreeDir(tree, joinPathFragments(dir, subDir)),
      );
  }
};
