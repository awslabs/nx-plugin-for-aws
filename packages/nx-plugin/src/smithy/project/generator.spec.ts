/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { readJson, Tree } from '@nx/devkit';
import {
  smithyProjectGenerator,
  SMITHY_PROJECT_GENERATOR_INFO,
} from './generator';
import { createTreeUsingTsSolutionSetup } from '../../utils/test';
import { sharedConstructsGenerator } from '../../utils/shared-constructs';
import { expectHasMetricTags } from '../../utils/metrics.spec';

describe('smithyProjectGenerator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
  });

  it('should generate smithy project with default options', async () => {
    await smithyProjectGenerator(tree, {
      name: 'test-api',
    });

    // Verify directory structure
    expect(tree.exists('test-api/src/main.smithy')).toBeTruthy();
    expect(tree.exists('test-api/src/operations/echo.smithy')).toBeTruthy();
    expect(tree.exists('test-api/build.Dockerfile')).toBeTruthy();
    expect(tree.exists('test-api/smithy-build.json')).toBeTruthy();
    expect(tree.exists('test-api/project.json')).toBeTruthy();

    // Verify project configuration
    const projectConfig = readJson(tree, 'test-api/project.json');
    expect(projectConfig.name).toBe('@proj/test-api');
    expect(projectConfig.projectType).toBe('library');
    expect(projectConfig.sourceRoot).toBe('test-api/src');
    expect(projectConfig.targets.build).toBeDefined();
    expect(projectConfig.targets.compile).toBeDefined();

    // Verify compile target configuration
    expect(projectConfig.targets.compile.executor).toBe('nx:run-commands');
    expect(projectConfig.targets.compile.options.commands).toEqual([
      'rimraf dist/test-api/build',
      'make-dir dist/test-api/build',
      'docker build -f test-api/build.Dockerfile --target export --output type=local,dest=dist/test-api/build test-api',
    ]);
    expect(projectConfig.targets.compile.outputs).toEqual([
      '{workspaceRoot}/dist/{projectRoot}/build',
    ]);
    expect(projectConfig.targets.build.dependsOn).toEqual(['compile']);

    // Create snapshots of generated files
    expect(tree.read('test-api/src/main.smithy', 'utf-8')).toMatchSnapshot(
      'main.smithy',
    );
    expect(
      tree.read('test-api/src/operations/echo.smithy', 'utf-8'),
    ).toMatchSnapshot('echo.smithy');
    expect(tree.read('test-api/build.Dockerfile', 'utf-8')).toMatchSnapshot(
      'build.Dockerfile',
    );
    expect(tree.read('test-api/smithy-build.json', 'utf-8')).toMatchSnapshot(
      'smithy-build.json',
    );
    expect(tree.read('test-api/project.json', 'utf-8')).toMatchSnapshot(
      'project.json',
    );
  });

  it('should generate smithy project with custom service name', async () => {
    await smithyProjectGenerator(tree, {
      name: 'test-api',
      serviceName: 'CustomService',
    });

    const mainSmithy = tree.read('test-api/src/main.smithy', 'utf-8');
    expect(mainSmithy).toContain('service CustomService');
    expect(mainSmithy).toContain('@title("CustomService")');
    expect(mainSmithy).toMatchSnapshot('custom-service-main.smithy');
  });

  it('should generate smithy project with custom namespace', async () => {
    await smithyProjectGenerator(tree, {
      name: 'test-api',
      namespace: 'com.example.custom',
    });

    const mainSmithy = tree.read('test-api/src/main.smithy', 'utf-8');
    expect(mainSmithy).toContain('namespace com.example.custom');
    expect(mainSmithy).toMatchSnapshot('custom-namespace-main.smithy');
  });

  it('should generate smithy project with custom directory', async () => {
    await smithyProjectGenerator(tree, {
      name: 'test-api',
      directory: 'apis',
    });

    // Verify directory structure
    expect(tree.exists('apis/test-api')).toBeTruthy();
    expect(tree.exists('apis/test-api/src')).toBeTruthy();
    expect(tree.exists('apis/test-api/src/main.smithy')).toBeTruthy();
    expect(tree.exists('apis/test-api/project.json')).toBeTruthy();

    // Verify project configuration
    const projectConfig = readJson(tree, 'apis/test-api/project.json');
    expect(projectConfig.sourceRoot).toBe('apis/test-api/src');
    expect(tree.read('apis/test-api/project.json', 'utf-8')).toMatchSnapshot(
      'custom-dir-project.json',
    );
  });

  it('should generate smithy project with subdirectory', async () => {
    await smithyProjectGenerator(tree, {
      name: 'test-api',
      directory: 'services',
      subDirectory: 'model',
    });

    // Verify directory structure
    expect(tree.exists('services/model')).toBeTruthy();
    expect(tree.exists('services/model/src')).toBeTruthy();
    expect(tree.exists('services/model/src/main.smithy')).toBeTruthy();
    expect(tree.exists('services/model/project.json')).toBeTruthy();

    // Verify project configuration
    const projectConfig = readJson(tree, 'services/model/project.json');
    expect(projectConfig.sourceRoot).toBe('services/model/src');
    expect(tree.read('services/model/project.json', 'utf-8')).toMatchSnapshot(
      'subdir-project.json',
    );
  });

  it('should generate smithy project with all custom options', async () => {
    await smithyProjectGenerator(tree, {
      name: 'test-api',
      serviceName: 'MyCustomService',
      namespace: 'com.mycompany.api',
      directory: 'backend',
      subDirectory: 'model',
    });

    // Verify directory structure
    expect(tree.exists('backend/model')).toBeTruthy();
    expect(tree.exists('backend/model/src/main.smithy')).toBeTruthy();

    const mainSmithy = tree.read('backend/model/src/main.smithy', 'utf-8');
    expect(mainSmithy).toContain('namespace com.mycompany.api');
    expect(mainSmithy).toContain('service MyCustomService');
    expect(mainSmithy).toContain('@title("MyCustomService")');

    const projectConfig = readJson(tree, 'backend/model/project.json');
    expect(projectConfig.sourceRoot).toBe('backend/model/src');

    expect(mainSmithy).toMatchSnapshot('all-custom-main.smithy');
    expect(tree.read('backend/model/project.json', 'utf-8')).toMatchSnapshot(
      'all-custom-project.json',
    );
  });

  it('should use npm scope for default namespace', async () => {
    // Set up npm scope in package.json
    const packageJson = readJson(tree, 'package.json');
    packageJson.name = '@myorg/workspace';
    tree.write('package.json', JSON.stringify(packageJson, null, 2));

    await smithyProjectGenerator(tree, {
      name: 'test-api',
    });

    const mainSmithy = tree.read('test-api/src/main.smithy', 'utf-8');
    expect(mainSmithy).toContain('namespace myorg');
    expect(mainSmithy).toMatchSnapshot('npm-scope-main.smithy');
  });

  it('should add generator metadata to project configuration', async () => {
    await smithyProjectGenerator(tree, {
      name: 'test-api',
    });

    const projectConfig = readJson(tree, 'test-api/project.json');
    expect(projectConfig.metadata).toHaveProperty(
      'generator',
      SMITHY_PROJECT_GENERATOR_INFO.id,
    );
    expect(projectConfig.metadata).toHaveProperty('apiName', 'test-api');
  });

  it('should add generator metric to app.ts when shared constructs exist', async () => {
    // Set up test tree with shared constructs
    await sharedConstructsGenerator(tree, { iacProvider: 'CDK' });

    // Call the generator function
    await smithyProjectGenerator(tree, {
      name: 'test-api',
    });

    // Verify the metric was added to app.ts
    expectHasMetricTags(tree, SMITHY_PROJECT_GENERATOR_INFO.metric);
  });

  it('should handle kebab-case conversion for service names', async () => {
    await smithyProjectGenerator(tree, {
      name: 'my-test-api',
      serviceName: 'MyTestService',
    });

    const mainSmithy = tree.read('my-test-api/src/main.smithy', 'utf-8');
    expect(mainSmithy).toContain('service MyTestService');
    expect(mainSmithy).toMatchSnapshot('kebab-case-main.smithy');

    const smithyBuild = tree.read('my-test-api/smithy-build.json', 'utf-8');
    expect(smithyBuild).toMatchSnapshot('kebab-case-smithy-build.json');
  });

  it('should generate valid Docker build configuration', async () => {
    await smithyProjectGenerator(tree, {
      name: 'test-api',
    });

    const dockerfile = tree.read('test-api/build.Dockerfile', 'utf-8');
    expect(dockerfile).toContain(
      'FROM public.ecr.aws/docker/library/node:24 AS builder',
    );
    expect(dockerfile).toMatchSnapshot('dockerfile');

    const projectConfig = readJson(tree, 'test-api/project.json');
    const dockerCommand = projectConfig.targets.compile.options.commands[2];
    expect(dockerCommand).toBe(
      'docker build -f test-api/build.Dockerfile --target export --output type=local,dest=dist/test-api/build test-api',
    );
  });

  it('should configure proper build dependencies', async () => {
    await smithyProjectGenerator(tree, {
      name: 'test-api',
    });

    const projectConfig = readJson(tree, 'test-api/project.json');
    expect(projectConfig.targets.build.dependsOn).toContain('compile');
    expect(projectConfig.targets.compile.cache).toBe(true);
  });

  it('should handle empty service name by using project name', async () => {
    await smithyProjectGenerator(tree, {
      name: 'my-service',
      serviceName: undefined,
    });

    const mainSmithy = tree.read('my-service/src/main.smithy', 'utf-8');
    expect(mainSmithy).toContain('service MyService');
    expect(mainSmithy).toContain('@title("MyService")');
    expect(mainSmithy).toMatchSnapshot('default-service-name-main.smithy');
  });
});
