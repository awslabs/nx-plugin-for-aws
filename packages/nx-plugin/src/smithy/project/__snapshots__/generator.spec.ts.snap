// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`smithyProjectGenerator > should generate smithy project with all custom options > all-custom-main.smithy 1`] = `
"$version: "2.0"

namespace com.mycompany.api

use aws.protocols#restJson1
use smithy.framework#ValidationException

/// TODO: describe your service here
@title("MyCustomService")
@restJson1
service MyCustomService {
    version: "1.0.0"
    operations: [
        Echo
    ]
    errors: [
        ValidationException
    ]
}
"
`;

exports[`smithyProjectGenerator > should generate smithy project with all custom options > all-custom-project.json 1`] = `
"{
  "name": "@proj/test-api",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "backend/model/src",
  "projectType": "library",
  "targets": {
    "build": {
      "dependsOn": ["compile"]
    },
    "compile": {
      "cache": true,
      "outputs": ["{workspaceRoot}/dist/{projectRoot}/build"],
      "executor": "nx:run-commands",
      "options": {
        "commands": [
          "rimraf dist/backend/model/build",
          "make-dir dist/backend/model/build",
          "docker build -f backend/model/build.Dockerfile --target export --output type=local,dest=dist/backend/model/build backend/model"
        ],
        "parallel": false,
        "cwd": "{workspaceRoot}"
      }
    }
  },
  "metadata": {
    "generator": "smithy#project",
    "apiName": "test-api"
  }
}
"
`;

exports[`smithyProjectGenerator > should generate smithy project with custom directory > custom-dir-project.json 1`] = `
"{
  "name": "@proj/test-api",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "apis/test-api/src",
  "projectType": "library",
  "targets": {
    "build": {
      "dependsOn": ["compile"]
    },
    "compile": {
      "cache": true,
      "outputs": ["{workspaceRoot}/dist/{projectRoot}/build"],
      "executor": "nx:run-commands",
      "options": {
        "commands": [
          "rimraf dist/apis/test-api/build",
          "make-dir dist/apis/test-api/build",
          "docker build -f apis/test-api/build.Dockerfile --target export --output type=local,dest=dist/apis/test-api/build apis/test-api"
        ],
        "parallel": false,
        "cwd": "{workspaceRoot}"
      }
    }
  },
  "metadata": {
    "generator": "smithy#project",
    "apiName": "test-api"
  }
}
"
`;

exports[`smithyProjectGenerator > should generate smithy project with custom namespace > custom-namespace-main.smithy 1`] = `
"$version: "2.0"

namespace com.example.custom

use aws.protocols#restJson1
use smithy.framework#ValidationException

/// TODO: describe your service here
@title("TestApi")
@restJson1
service TestApi {
    version: "1.0.0"
    operations: [
        Echo
    ]
    errors: [
        ValidationException
    ]
}
"
`;

exports[`smithyProjectGenerator > should generate smithy project with custom service name > custom-service-main.smithy 1`] = `
"$version: "2.0"

namespace proj

use aws.protocols#restJson1
use smithy.framework#ValidationException

/// TODO: describe your service here
@title("CustomService")
@restJson1
service CustomService {
    version: "1.0.0"
    operations: [
        Echo
    ]
    errors: [
        ValidationException
    ]
}
"
`;

exports[`smithyProjectGenerator > should generate smithy project with default options > build.Dockerfile 1`] = `
"FROM public.ecr.aws/docker/library/node:24 AS builder

# Output directory
RUN mkdir /out

# Install Smithy CLI
# https://smithy.io/2.0/guides/smithy-cli/cli_installation.html
WORKDIR /smithy
ARG TARGETPLATFORM
RUN if [ "$TARGETPLATFORM" = "linux/arm64" ]; then ARCH="aarch64"; else ARCH="x86_64"; fi && \\
    mkdir -p smithy-install/smithy && \\
    curl -L https://github.com/smithy-lang/smithy/releases/download/1.61.0/smithy-cli-linux-$ARCH.zip -o smithy-install/smithy-cli-linux-$ARCH.zip && \\
    unzip -qo smithy-install/smithy-cli-linux-$ARCH.zip -d smithy-install && \\
    mv smithy-install/smithy-cli-linux-$ARCH/* smithy-install/smithy
RUN smithy-install/smithy/install

# Add node dependencies for bundling
WORKDIR /project
RUN npm i -g pnpm@10.15.1 rolldown@1.0.0-beta.38

# Copy project files
COPY smithy-build.json .
COPY src src

# Smithy build with Maven cache mount
RUN --mount=type=cache,target=/root/.m2/repository,id=maven-cache \\
    smithy build

# Copy OpenAPI specification to output location
RUN mkdir -p /out/openapi
RUN cp /project/build/smithy/source/openapi/*.openapi.json /out/openapi/openapi.json

WORKDIR /project/build/smithy/source/typescript-ssdk-codegen

# Install SSDK dependencies with pnpm cache mount
RUN --mount=type=cache,target=/root/.local/share/pnpm/store,id=pnpm-store \\
    --mount=type=cache,target=/project/build/smithy/source/typescript-ssdk-codegen/node_modules,id=ssdk-node-modules \\
    pnpm install --prefer-offline

# Install rolldown plugins with pnpm cache mount
RUN --mount=type=cache,target=/root/.local/share/pnpm/store,id=pnpm-store \\
    --mount=type=cache,target=/project/build/smithy/source/typescript-ssdk-codegen/node_modules,id=ssdk-node-modules \\
    pnpm add -D rolldown-plugin-dts@0.16.5 @rollup/plugin-esm-shim@0.1.8

RUN cat <<EOF > rolldown.config.js
import { dts } from 'rolldown-plugin-dts';
import esmShim from '@rollup/plugin-esm-shim';
import fs from 'fs';
import path from 'path';

// Bundle the re2-wasm wasm file as part of the module
const re2WasmPlugin = () => ({
  name: 're2-wasm-plugin',
  resolveId: (id) => {
    if (!(id === 're2-wasm' || id.endsWith('/re2-wasm'))) return null;
    try {
      const dir = path.join(path.dirname(require.resolve('re2-wasm/package.json')), 'build/wasm');
      return fs.existsSync(path.join(dir, 're2.wasm')) && fs.existsSync(path.join(dir, 're2.js')) ? path.join(dir, 're2.js') : null;
    } catch { return null; }
  },
  load: (id) => {
    if (!(id.endsWith('re2.js') && id.includes('re2-wasm'))) return null;
    try {
      const wasmPath = path.join(path.dirname(id), 're2.wasm');
      return fs.existsSync(wasmPath) && fs.existsSync(id) ? \\\`var Module = { wasmBinary: Buffer.from("\\\${fs.readFileSync(wasmPath).toString('base64')}", "base64") }\\n\\\${fs.readFileSync(id, 'utf8')}\\\` : null;
    } catch { return null; }
  }
});

export default {
  input: './src/index.ts',
  plugins: [
    re2WasmPlugin(),
    dts({ resolve: true }),
    esmShim(),
  ],
  output: [{
    dir: 'dist',
    format: 'es',
  }],
  resolve: {
    tsconfigFilename: 'tsconfig.es.json',
  },
  platform: 'node',
};
EOF

# Create SSDK bundle with rolldown
RUN --mount=type=cache,target=/project/build/smithy/source/typescript-ssdk-codegen/node_modules,id=ssdk-node-modules \\
    rolldown -c

RUN mkdir -p /out/ssdk
RUN cp dist/* /out/ssdk/

# Export the /out directory
FROM scratch AS export
COPY --from=builder /out /
"
`;

exports[`smithyProjectGenerator > should generate smithy project with default options > echo.smithy 1`] = `
"$version: "2.0"

namespace proj

/// An example operation
@http(method: "GET", uri: "/echo")
@readonly
operation Echo {
    input := {
        @httpQuery("message")
        @required
        message: String
    }
    output := {
        @required
        message: String
    }
}
"
`;

exports[`smithyProjectGenerator > should generate smithy project with default options > main.smithy 1`] = `
"$version: "2.0"

namespace proj

use aws.protocols#restJson1
use smithy.framework#ValidationException

/// TODO: describe your service here
@title("TestApi")
@restJson1
service TestApi {
    version: "1.0.0"
    operations: [
        Echo
    ]
    errors: [
        ValidationException
    ]
}
"
`;

exports[`smithyProjectGenerator > should generate smithy project with default options > project.json 1`] = `
"{
  "name": "@proj/test-api",
  "$schema": "../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "test-api/src",
  "projectType": "library",
  "targets": {
    "build": {
      "dependsOn": ["compile"]
    },
    "compile": {
      "cache": true,
      "outputs": ["{workspaceRoot}/dist/{projectRoot}/build"],
      "executor": "nx:run-commands",
      "options": {
        "commands": [
          "rimraf dist/test-api/build",
          "make-dir dist/test-api/build",
          "docker build -f test-api/build.Dockerfile --target export --output type=local,dest=dist/test-api/build test-api"
        ],
        "parallel": false,
        "cwd": "{workspaceRoot}"
      }
    }
  },
  "metadata": {
    "generator": "smithy#project",
    "apiName": "test-api"
  }
}
"
`;

exports[`smithyProjectGenerator > should generate smithy project with default options > smithy-build.json 1`] = `
"{
  "version": "1.0",
  "sources": ["src/"],
  "plugins": {
    "openapi": {
      "service": "proj#TestApi",
      "version": "3.1.0",
      "tags": true,
      "useIntegerType": true
    },
    "typescript-ssdk-codegen": {
      "package": "@proj/test-api-ssdk",
      "packageVersion": "0.0.1"
    }
  },
  "maven": {
    "dependencies": [
      "software.amazon.smithy:smithy-model:1.61.0",
      "software.amazon.smithy:smithy-aws-traits:1.61.0",
      "software.amazon.smithy:smithy-validation-model:1.61.0",
      "software.amazon.smithy:smithy-openapi:1.61.0",
      "software.amazon.smithy.typescript:smithy-aws-typescript-codegen:0.34.1"
    ]
  }
}
"
`;

exports[`smithyProjectGenerator > should generate smithy project with subdirectory > subdir-project.json 1`] = `
"{
  "name": "@proj/test-api",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "services/model/src",
  "projectType": "library",
  "targets": {
    "build": {
      "dependsOn": ["compile"]
    },
    "compile": {
      "cache": true,
      "outputs": ["{workspaceRoot}/dist/{projectRoot}/build"],
      "executor": "nx:run-commands",
      "options": {
        "commands": [
          "rimraf dist/services/model/build",
          "make-dir dist/services/model/build",
          "docker build -f services/model/build.Dockerfile --target export --output type=local,dest=dist/services/model/build services/model"
        ],
        "parallel": false,
        "cwd": "{workspaceRoot}"
      }
    }
  },
  "metadata": {
    "generator": "smithy#project",
    "apiName": "test-api"
  }
}
"
`;

exports[`smithyProjectGenerator > should generate valid Docker build configuration > dockerfile 1`] = `
"FROM public.ecr.aws/docker/library/node:24 AS builder

# Output directory
RUN mkdir /out

# Install Smithy CLI
# https://smithy.io/2.0/guides/smithy-cli/cli_installation.html
WORKDIR /smithy
ARG TARGETPLATFORM
RUN if [ "$TARGETPLATFORM" = "linux/arm64" ]; then ARCH="aarch64"; else ARCH="x86_64"; fi && \\
    mkdir -p smithy-install/smithy && \\
    curl -L https://github.com/smithy-lang/smithy/releases/download/1.61.0/smithy-cli-linux-$ARCH.zip -o smithy-install/smithy-cli-linux-$ARCH.zip && \\
    unzip -qo smithy-install/smithy-cli-linux-$ARCH.zip -d smithy-install && \\
    mv smithy-install/smithy-cli-linux-$ARCH/* smithy-install/smithy
RUN smithy-install/smithy/install

# Add node dependencies for bundling
WORKDIR /project
RUN npm i -g pnpm@10.15.1 rolldown@1.0.0-beta.38

# Copy project files
COPY smithy-build.json .
COPY src src

# Smithy build with Maven cache mount
RUN --mount=type=cache,target=/root/.m2/repository,id=maven-cache \\
    smithy build

# Copy OpenAPI specification to output location
RUN mkdir -p /out/openapi
RUN cp /project/build/smithy/source/openapi/*.openapi.json /out/openapi/openapi.json

WORKDIR /project/build/smithy/source/typescript-ssdk-codegen

# Install SSDK dependencies with pnpm cache mount
RUN --mount=type=cache,target=/root/.local/share/pnpm/store,id=pnpm-store \\
    --mount=type=cache,target=/project/build/smithy/source/typescript-ssdk-codegen/node_modules,id=ssdk-node-modules \\
    pnpm install --prefer-offline

# Install rolldown plugins with pnpm cache mount
RUN --mount=type=cache,target=/root/.local/share/pnpm/store,id=pnpm-store \\
    --mount=type=cache,target=/project/build/smithy/source/typescript-ssdk-codegen/node_modules,id=ssdk-node-modules \\
    pnpm add -D rolldown-plugin-dts@0.16.5 @rollup/plugin-esm-shim@0.1.8

RUN cat <<EOF > rolldown.config.js
import { dts } from 'rolldown-plugin-dts';
import esmShim from '@rollup/plugin-esm-shim';
import fs from 'fs';
import path from 'path';

// Bundle the re2-wasm wasm file as part of the module
const re2WasmPlugin = () => ({
  name: 're2-wasm-plugin',
  resolveId: (id) => {
    if (!(id === 're2-wasm' || id.endsWith('/re2-wasm'))) return null;
    try {
      const dir = path.join(path.dirname(require.resolve('re2-wasm/package.json')), 'build/wasm');
      return fs.existsSync(path.join(dir, 're2.wasm')) && fs.existsSync(path.join(dir, 're2.js')) ? path.join(dir, 're2.js') : null;
    } catch { return null; }
  },
  load: (id) => {
    if (!(id.endsWith('re2.js') && id.includes('re2-wasm'))) return null;
    try {
      const wasmPath = path.join(path.dirname(id), 're2.wasm');
      return fs.existsSync(wasmPath) && fs.existsSync(id) ? \\\`var Module = { wasmBinary: Buffer.from("\\\${fs.readFileSync(wasmPath).toString('base64')}", "base64") }\\n\\\${fs.readFileSync(id, 'utf8')}\\\` : null;
    } catch { return null; }
  }
});

export default {
  input: './src/index.ts',
  plugins: [
    re2WasmPlugin(),
    dts({ resolve: true }),
    esmShim(),
  ],
  output: [{
    dir: 'dist',
    format: 'es',
  }],
  resolve: {
    tsconfigFilename: 'tsconfig.es.json',
  },
  platform: 'node',
};
EOF

# Create SSDK bundle with rolldown
RUN --mount=type=cache,target=/project/build/smithy/source/typescript-ssdk-codegen/node_modules,id=ssdk-node-modules \\
    rolldown -c

RUN mkdir -p /out/ssdk
RUN cp dist/* /out/ssdk/

# Export the /out directory
FROM scratch AS export
COPY --from=builder /out /
"
`;

exports[`smithyProjectGenerator > should handle empty service name by using project name > default-service-name-main.smithy 1`] = `
"$version: "2.0"

namespace proj

use aws.protocols#restJson1
use smithy.framework#ValidationException

/// TODO: describe your service here
@title("MyService")
@restJson1
service MyService {
    version: "1.0.0"
    operations: [
        Echo
    ]
    errors: [
        ValidationException
    ]
}
"
`;

exports[`smithyProjectGenerator > should handle kebab-case conversion for service names > kebab-case-main.smithy 1`] = `
"$version: "2.0"

namespace proj

use aws.protocols#restJson1
use smithy.framework#ValidationException

/// TODO: describe your service here
@title("MyTestService")
@restJson1
service MyTestService {
    version: "1.0.0"
    operations: [
        Echo
    ]
    errors: [
        ValidationException
    ]
}
"
`;

exports[`smithyProjectGenerator > should handle kebab-case conversion for service names > kebab-case-smithy-build.json 1`] = `
"{
  "version": "1.0",
  "sources": ["src/"],
  "plugins": {
    "openapi": {
      "service": "proj#MyTestService",
      "version": "3.1.0",
      "tags": true,
      "useIntegerType": true
    },
    "typescript-ssdk-codegen": {
      "package": "@proj/my-test-service-ssdk",
      "packageVersion": "0.0.1"
    }
  },
  "maven": {
    "dependencies": [
      "software.amazon.smithy:smithy-model:1.61.0",
      "software.amazon.smithy:smithy-aws-traits:1.61.0",
      "software.amazon.smithy:smithy-validation-model:1.61.0",
      "software.amazon.smithy:smithy-openapi:1.61.0",
      "software.amazon.smithy.typescript:smithy-aws-typescript-codegen:0.34.1"
    ]
  }
}
"
`;

exports[`smithyProjectGenerator > should use npm scope for default namespace > npm-scope-main.smithy 1`] = `
"$version: "2.0"

namespace myorg

use aws.protocols#restJson1
use smithy.framework#ValidationException

/// TODO: describe your service here
@title("TestApi")
@restJson1
service TestApi {
    version: "1.0.0"
    operations: [
        Echo
    ]
    errors: [
        ValidationException
    ]
}
"
`;
